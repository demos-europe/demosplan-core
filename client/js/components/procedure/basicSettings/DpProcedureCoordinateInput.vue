<license>
  (c) 2010-present DEMOS plan GmbH.

  This file is part of the package demosplan,
  for more information see the license file.

  All rights reserved
</license>

<template>
  <div class="block u-mb-0_5 u-pv-0_5 border--top cf">
    <label
      class="inline-block u-m-0"
      for="customLatitude">
      {{ Translator.trans('coordinate.right.value') }}
    </label>
    <input
      type="text"
      pattern="[0-9]*[.|,]?[0-9]+"
      required
      id="customLatitude"
      v-model="latitudeValue"
      class="c-ol-map__select width-120 u-mr">

    <label
      class="inline-block u-m-0"
      for="customLongitude">
      {{ Translator.trans('coordinate.top.value') }}
    </label>
    <input
      type="text"
      pattern="[0-9]*[.|,]?[0-9]+"
      required
      id="customLongitude"
      v-model="longitudeValue"
      class="c-ol-map__select width-120 u-mr">

    <button
      @click.prevent="addMarker"
      :disabled="!isCoordinatesValid"
      ref="myBtnCoordinates"
      class="btn btn--primary float--right">
      {{ Translator.trans('coordinate.location.submite') }}
    </button>
  </div>
</template>

<script>

import proj4 from 'proj4'

export default {
  name: 'DpProcedureCoordinateInput',

  props: {
    coordinate: {
      required: false,
      type: Array,
      default: () => []
    }
  },

  data () {
    return {
      latitudeValue: '', // 568400.97
      longitudeValue: '' // 5923963.03
    }
  },

  computed: {
    isCoordinatesValid () {
      const lat = this.convertToFloat(this.latitudeValue)
      const lon = this.convertToFloat(this.longitudeValue)
      return Number(lat) === lat && Number(lon) === lon
    }
  },

  watch: {
    coordinate (coordinates) {
      this.updateCoordinates(coordinates)
    }
  },

  methods: {
    addMarker () {
      // Coordinate input field is meant to be EPSG:25832, therefore it needs to be transformed
      proj4.defs([
        [
          'EPSG:25832',
          '+proj=utm +zone=32 +ellps=GRS80 +units=m +no_defs'
        ]
      ])

      this.$emit('input', proj4('EPSG:25832', window.dplan.defaultProjectionLabel, [this.convertToFloat(this.latitudeValue), this.convertToFloat(this.longitudeValue)]))
      this.$refs.myBtnCoordinates.blur()
    },

    updateCoordinates (coordinates) {
      if (coordinates.length === 2) {
        this.latitudeValue = coordinates[0]
        this.longitudeValue = coordinates[1]
      }
    },

    convertToFloat (val) {
      return parseFloat((val + '').replace(',', '.'))
    }
  },

  mounted () {
    //  Setup state + behavior
    this.updateCoordinates(this.coordinate)
  }
}
</script>
