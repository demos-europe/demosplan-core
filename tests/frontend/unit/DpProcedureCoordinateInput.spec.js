/**
 * (c) 2010-present DEMOS E-Partizipation GmbH.
 *
 * This file is part of the package demosplan,
 * for more information see the license file.
 *
 * All rights reserved
 */

import DpProcedureCoordinateInput from '@DpJs/components/procedure/basicSettings/DpProcedureCoordinateInput'
import shallowMountWithGlobalMocks from '@DpJs/VueConfigLocal'

describe('DpProcedureCoordinateInput', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMountWithGlobalMocks(DpProcedureCoordinateInput, {})
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should check if coordinates are valid', () => {
    wrapper.setData({ latitudeValue: '123', longitudeValue: '456' })
    expect(wrapper.vm.isCoordinatesValid).toBe(true)

    wrapper.setData({ latitudeValue: 'abc', longitudeValue: '456' })
    expect(wrapper.vm.isCoordinatesValid).toBe(false)
  })

  it('should update coordinates correctly', () => {
    wrapper.vm.updateCoordinates(['789', '012'])
    expect(wrapper.vm.latitudeValue).toBe('789')
    expect(wrapper.vm.longitudeValue).toBe('012')
  })

  it('should convert string to float correctly', () => {
    expect(wrapper.vm.convertToFloat('123.456')).toBe(123.456)
    expect(wrapper.vm.convertToFloat('123,456')).toBe(123.456)
  })

  it('should set coordinate prop and check types of latitudeValue and longitudeValue', async () => {
    await wrapper.setProps({ coordinate: ['123', '456'] })
    expect(typeof wrapper.vm.latitudeValue).toBe('string')
    expect(typeof wrapper.vm.longitudeValue).toBe('string')
    expect(typeof wrapper.vm.convertToFloat(wrapper.vm.latitudeValue)).toBe('number')
    expect(typeof wrapper.vm.convertToFloat(wrapper.vm.longitudeValue)).toBe('number')
  })

  it('should enable the button with valid and disable with invalid input from props', async () => {
    await wrapper.setProps({ coordinate: ['123', '456'] })
    expect(wrapper.html()).not.toContain('disabled="disabled"')

    await wrapper.setProps({ coordinate: ['abc', '456'] })
    expect(wrapper.html()).toContain('disabled="disabled"')
  })
})
