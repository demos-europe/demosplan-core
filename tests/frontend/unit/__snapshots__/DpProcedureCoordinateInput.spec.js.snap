// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DpProcedureCoordinateInput should match the snapshot 1`] = `"<div class="block u-mb-0_5 u-pv-0_5 border--top flow-root"><label class="inline-block u-m-0" for="customLatitude">coordinate.right.value</label><input id="customLatitude" type="text" pattern="[0-9]*[.|,]?[0-9]+" required="" class="c-ol-map__select w-9 u-mr"><label class="inline-block u-m-0" for="customLongitude">coordinate.top.value</label><input id="customLongitude" type="text" pattern="[0-9]*[.|,]?[0-9]+" required="" class="c-ol-map__select w-9 u-mr"><button disabled="" class="btn btn--primary float-right">coordinate.location.submite</button></div>"`;
